libcob/cconv.c

```diff
@@ -20,454 +20,142 @@
-	0xC8, 0xC9, 0xD1, 0xD2, 0xD3, 0xD4, 0xD5, 0xD6,
-	0xD7, 0xD8, 0xD9, 0xE2, 0xE3, 0xE4, 0xE5, 0xE6,
-	0xE7, 0xE8, 0xE9, 0x4A, 0xE0, 0x5A, 0x5F, 0x6D,
-	0x79, 0x81, 0x82, 0x83, 0x84, 0x85, 0x86, 0x87,
-	0x88, 0x89, 0x91, 0x92, 0x93, 0x94, 0x95, 0x96,
-	0x97, 0x98, 0x99, 0xA2, 0xA3, 0xA4, 0xA5, 0xA6,
-	0xA7, 0xA8, 0xA9, 0xC0, 0x6A, 0xD0, 0xA1, 0x07,
-	0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F,
-	0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F,
-	0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F,
-	0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F,
-	0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F,
-	0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F,
-	0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F,
-	0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F,
-	0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F,
-	0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F,
-	0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F,
-	0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F,
-	0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F,
-	0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F,
-	0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F,
-	0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F, 0x3F
-};
+static int
+cob_convert_hex_digit (char h)
+{
+	if (h >= '0' && h <= '9') return h - '0';
+	else if (h >= 'A' && h <= 'F') return 10 + h - 'A';
+	else if (h >= 'a' && h <= 'f') return 10 + h - 'a';
+	else return -1;
+}
 
-const char *
-cob_get_collation_name (int col_id)
+static int
+cob_convert_hex_byte (const char *h)
 {
-	switch (col_id) {
-	case CB_EBCDIC_DEFAULT:
-		return "DEFAULT";
-	case CB_EBCDIC_RESTRICTED_GC:
-		return "RESTRICTED-GC";
-	case CB_EBCDIC_IBM:
-		return "IBM";
-	case CB_EBCDIC_GCOS:
-		return "GCOS";
-	default:
-		return NULL;
+	int d1 = cob_convert_hex_digit (h[0]);
+	int d2 = cob_convert_hex_digit (h[1]);
+	if (d1 < 0 || d2 < 0) {
+		return -1;
+	} else {
+		return d1 * 16 + d2;
 	}
 }
 
-static int
-cob_get_collation_by_id (int col_id,
-			 const cob_u8_t **p_ebcdic_as_ascii,
-			 const cob_u8_t **p_ascii_as_ebcdic)
+static char *
+cob_skip_blanks(const char *s)
 {
-	const cob_u8_t *ebcdic_as_ascii;
-	const cob_u8_t *ascii_as_ebcdic;
+  while (*s != '\0' && (isblank(*s) || *s == '\n')) {
+    ++s;
+  }
+  return s;
+}
+
+/* Note: cobc uses this function to create tables that are embedded in the
+   generated code, and we also make it available in libcob so that external
+   tools that make use of the library can use this feature */
+int
+cob_load_collation (const char *col_name,
+		    cob_u8_t *ebcdic_to_ascii,
+		    cob_u8_t *ascii_to_ebcdic)
+{
+	cob_u8_t table[512];
+	char hex[COB_SMALL_BUFF];
+        char *hexptr;
+	int i, n, line;
+	FILE *f;
+        char *config_dir;
+	char filename[COB_FILE_BUFF];
+	const char *last_col_name = NULL;
+
+	if (col_name[0] == '.' || col_name[0] == SLASH_CHAR
+#ifdef _WIN32
+ || col_name[0] != '\0' && col_name[1] == ':'
+#endif
+	   ) {
+		/* If it's a path, use it as-is */
+		snprintf (filename, (size_t)COB_FILE_MAX, "%s", col_name);
```

Comment:
please `strlen` + check sizeof (filename) + memcpy instead

Reply:
Alright. Is there a specific error to generate to indicate the string would overflow ? Or should I just truncate ?

Reply:
Alright. And for the second case (`snprintf` with `COB_CONFIG_DIR` and so on), should I switch to memcpy ? Or just checking the length is enough ?

Reply:
length check is enough